opencxl/cxl/device/cxl_type1_device.py

```diff
@@ -201,11 +219,92 @@ def _init_device(
     def get_reg_vals(self):
         return self._cxl_io_manager.get_cfg_reg_vals()
 
-    async def cxl_cache_readline(self, addr: int) -> int:
-        pass
+    async def get_next_cqid(self) -> int:
+        cqid: int
+        async with self._cqid_assign_lock:
+            cqid = next(self._cqid_gen)
+        return cqid
+
+    async def cxl_cache_readline(self, addr: int, cqid: Optional[int] = None) -> Future[int]:
+        loop = get_running_loop()
+        fut_data = loop.create_future()  # ugly solution
+
+        def _listen_check_set_fut(pckt: CxlCacheH2DRspPacket | CxlCacheH2DDataPacket):
+            # for now, ignore the 32B transfer case.
+            if isinstance(pckt, CxlCacheH2DDataPacket):
+                # received a data packet
+                # since we're ignoring the 32B transfer case,
+                # we can just assume this data packet contains everything we want
+                fut_data.set_result(pckt.data)
+                raise CancelledError
```

Comment:
`# we can just assume this data packet contains everything we want`, but this throws an exception (and being called outside of a `try` block). Is this an error that should crash? If so, let's print an error message and crash. Otherwise, let it keep running.

Reply:
Raising CancelledError cancels the calling task. It's like pthread_exit(). I don't really like the design of this approach, but I couldn't really think of a better solution.

Reply:
In principle, it is bad practice to raise an exception without catching it unless you are trying to crash the system due to an error. If this "error" does not actually crash the system, that's even more concerning given what impact would it have to the callers?

Please take a look at the following page and see how it could be applied to your current code.
https://www.slingacademy.com/article/python-handling-asyncio-exceptions-cancellederror-gracefully